### ISSUE TITLE
search: allow queries to contain \n (special case of multiline search)

### ISSUE THREAD

**Issue by ijt** [sourcegraph/sourcegraph#4138]
This is a subset of the functionality proposed in #35, but easier to implement in a performant way.

**Comment by slimsag** [comment]
The idea here is that queries today have a structure like the following EBNF grammar ([explanation](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form#Table_of_symbols)):

```ebnf
LineMatch = Regexp | QuotedStringLiteral ;
Query     = [ SearchFilters ] , LineMatch , [ SearchFilters ] ;
```

This is to say that you can specify search filters on either side of your regexp or `"quoted string literal"` search, and we'll find any lines in files that contain that `LineMatch`, up to the newline.

After this change, you would be able to specify any number of newlines and `LineMatch` to search across multiple consecutive lines:

```ebnf
LineMatch = Regexp | QuotedStringLiteral ;
Query     = { [ SearchFilters ] , LineMatch , [ SearchFilters ] , NewlineMatch } ;
```

Example:

```
package main

func main() {
    fmt.Println("foobar")
}
```

- Query `package .*\n\n"func main() {"` would match lines 1 (`package .*\n`), 2 (`\n`), and 3 (`"func main() {"`) above.
- Query `func .*\(\)\s\{\n.*\n}` would match lines 3 (`func .*\(\)\s{\n`), 4 (`.*\n`), and 5 (`}`).

In other words, this allows searching over multiple consecutive lines.

What it doesn't allow (which is harder to do, and what https://github.com/sourcegraph/sourcegraph/issues/35 would allow for) us full-blown multi line search where you could write: `package main.*}` and match the entire file above (`.*` matching multiple lines).

**Comment by slimsag** [comment]
There are some further examples in the third item of https://github.com/sourcegraph/infrastructure/issues/1475#issuecomment-493611685 which may be more understandable than the above specifics at a high level.

Also, as mentioned in https://github.com/sourcegraph/infrastructure/issues/1475#issuecomment-495021234 for the $CUSTOMER in question this would be intrinsically dependent on also having https://github.com/sourcegraph/sourcegraph/issues/4135
